context/context.go

```diff
@@ -0,0 +1,141 @@
+package context
+
+import (
+	gocontext "context"
+	"fmt"
+
+	"github.com/flanksource/commons/logger"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type ContextOptions func(*Context)
+
+func WithTraceFn(fn func(*Context) bool) ContextOptions {
+	return func(opts *Context) {
+		opts.isTraceFn = fn
+	}
+}
+
+func WithDebugFn(fn func(*Context) bool) ContextOptions {
+	return func(opts *Context) {
+		opts.isDebugFn = fn
+	}
+}
+
+func WithTracer(tracer trace.Tracer) ContextOptions {
+	return func(opts *Context) {
+		opts.tracer = tracer
+	}
+}
+
+func WithLogger(log logger.Logger) ContextOptions {
+	return func(opts *Context) {
+		opts.logger = log
+	}
+}
+
+func NewContext(opts ...ContextOptions) *Context {
+	ctx := Context{
+		Context: gocontext.Background(),
+	}
+	for _, opt := range opts {
+		opt(&ctx)
+	}
+	if ctx.logger == nil {
+		ctx.logger = logger.StandardLogger()
+	}
+	if ctx.isDebugFn == nil {
+		ctx.isDebugFn = func(*Context) bool {
+			return ctx.logger.IsDebugEnabled()
+		}
+	}
+	if ctx.isTraceFn == nil {
+		ctx.isTraceFn = func(*Context) bool {
+			return ctx.logger.IsTraceEnabled()
+		}
+	}
+	if ctx.tracer == nil {
+		ctx.tracer = trace.NewNoopTracerProvider().Tracer("noop")
+	}
+	return &ctx
+}
+
+type Context struct {
+	gocontext.Context
+	logger    logger.Logger
+	isDebugFn func(*Context) bool
+	isTraceFn func(*Context) bool
+	tracer    trace.Tracer
+}
+
+func (c *Context) WithTracer(tracer trace.Tracer) {
+	c.tracer = tracer
+}
+
+func (c *Context) WithValue(key, val interface{}) *Context {
+	return &Context{
+		Context:   gocontext.WithValue(c, key, val),
+		isDebugFn: c.isDebugFn,
+		logger:    c.logger,
+		isTraceFn: c.isTraceFn,
+		tracer:    c.tracer,
+	}
+}
+
+func (c *Context) IsDebug() bool {
+	return c.isDebugFn(c)
+}
+
+func (c *Context) IsTrace() bool {
+	return c.isTraceFn(c)
+}
+
+func (c *Context) Debugf(format string, args ...interface{}) {
+	if c.isDebugFn(c) {
+		if c.HasSpanStarted() {
+			c.GetSpan().AddEvent(fmt.Sprintf(format, args...), trace.WithAttributes(attribute.String("level", "debug")))
+		}
+		c.logger.Debugf(format, args...)
+	}
+}
+
+func (c *Context) Tracef(format string, args ...interface{}) {
+	if c.isTraceFn(c) {
+		if c.HasSpanStarted() {
+			c.GetSpan().AddEvent(fmt.Sprintf(format, args...), trace.WithAttributes(attribute.String("level", "trace")))
+		}
+		c.logger.Tracef(format, args...)
+	}
+}
```

Comment:
It feels like we're packing up everything inside context. I don't it it should be the one that's doing the logging.

`ctx.Debugf()` doesn't feel right.

Reply:
It does because ctx knows where (span, console) and if (logger levels, ctx.Object trace annotations etc..) to log at all, 

Otherwise that logic is spread out and not consistent e.g. if we wanted to trace logs for a specific user, or a specific job,